### PR TITLE
Tracking Fast cleaner during pattern recognition

### PR DESCRIPTION
**Pull request by VinInn**
This PR introduces a new implementation of the TrajectoryCleaner during pattern reco.
It is much faster of the previous one as it relies on the fact that all trajectories originate from the same seed.
More important it now counts degrees of freedom instead of hits to account for cases a mix of 1D,2D and 4D hits are used during pattern recognition.
It also allows not to count hits that do not contribute to the chi2 (such as seed-hits during in-out)

A small regression may be observed as the lostHitPenalty of the trajectory builder have been lowered to 20 (from 30) to make it consistent with the value used elsewhere.

For the time being it affects only reconstruction using the "GroupedCkfTrajectoryBuilder"
(so, no HLT, Muon )

In future we need to make a massive cleanup of all these "Cleaners"

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild , please test

File: TrackingTools/TrajectoryCleaning/interface/FastTrajectoryCleaner.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+
+  FastTrajectoryCleaner(const edm::ParameterSet & iConfig) :
+    validHitBonus_(0.5*iConfig.getParameter<double>("ValidHitBonus")),
+    missingHitPenalty_(iConfig.getParameter<double>("MissingHitPenalty")),
+    dismissSeed_(iConfig.getParameter<bool>("dismissSeed")){}
+
+  ~FastTrajectoryCleaner(){}
+
+  TrajectoryCleaner::clean;
```
Comment (davidlt) [inline TrackingTools/TrajectoryCleaning/interface/FastTrajectoryCleaner.h]: What is this line? Is it intended?

Comment (VinInn) [inline TrackingTools/TrajectoryCleaning/interface/FastTrajectoryCleaner.h]: interesting... gcc is happy about that??

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by VinInn** [comment]
```
cmsRun: /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-04-28-2300/src/TrackingTools/TrajectoryCleaning/src/FastTrajectoryCleaner.cc:31: virtual void FastTrajectoryCleaner::clean(FastTrajectoryCleaner::TempTrajectoryContainer&) const: Assertion `bestTr' failed.
```

interesting
1) how could be that GroupedCKf is called at HLT (by heavyIon?)
2) this points to a more fundamental issue: it arrives there with no valid candidates!

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by makortel** [comment]
@VinInn 

> 1) how could be that GroupedCKf is called at HLT (by heavyIon?)

Maybe GroupedCkf is coming from RECO (cmsDriver has `-s 
 HLT:HIon,RAW2DIGI,L1Reco,RECO`)?

**Comment by slava77** [comment]
@VinInn @makortel 
Please provide MTV plots as a reference for expected changes from this PR.
Thank you.

**Comment by VinInn** [comment]
I have some from a month ago
http://innocent.home.cern.ch/innocent/RelVal/pu35_81_FastClean/

Matti can provide some for PhaseI

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/14299	5c25610
- code changes are OK
- jenkins tests pass and comparisons with baseline show small differences starting from tracking variables, propagating to downstream variables without any significant impact on physics performance.
- higher stat tests on ttbar PU35 provided in http://innocent.home.cern.ch/innocent/RelVal/pu35_81_FastClean/ show no significant change
- local tests with additional workflows confirm the expected negligible impact on physics performance
- timing check on 70 events on 25202 (skipping the first event for initialization) shows a somewhat consistent reduction in CPU in track building by about 3% with a total of about 0.5% of the total RECO time reduction.
```
   -0.069945      -0.01%        91.33 ms/ev ->        85.16 ms/ev jetCoreRegionalStepTrackCandidates
   -0.039307      -0.03%       629.79 ms/ev ->       605.51 ms/ev initialStepTrackCandidates
   -0.036057      -0.03%       634.78 ms/ev ->       612.29 ms/ev initialStepTrackCandidatesPreSplitting
   -0.033658      -0.02%       579.64 ms/ev ->       560.46 ms/ev detachedTripletStepTrackCandidates
   -0.032069      -0.02%       460.39 ms/ev ->       445.86 ms/ev lowPtTripletStepTrackCandidates
   -0.023825      -0.02%       875.91 ms/ev ->       855.29 ms/ev pixelPairStepTrackCandidates
   -0.015689      -0.01%       411.71 ms/ev ->       405.30 ms/ev pixelLessStepTrackCandidates
   -0.011872      -0.00%       211.06 ms/ev ->       208.57 ms/ev tobTecStepTrackCandidates
   +0.010836      +0.00%       191.83 ms/ev ->       193.92 ms/ev convTrackCandidates
Total in printed:        4086.44 ==> 3972.36 d: -114.08(-0.0279167)
```

**Comment by davidlange6** [comment]
+1

**Comment by makortel** [comment]
@VinInn @slava77 For the record, here are the plots for phase1 (baseline being the tracking introduced in #14327)

**Comment by davidlange6** [comment]
this PR is apparently the cause of low PT changes (20% of tracks). Could you confirm that is indeed agreed (it was not mentioned in the discussion above) @VinInn

**Comment by slava77** [comment]
MTV plots show no real change vs pt
http://innocent.home.cern.ch/innocent/RelVal/pu35_81_FastClean/plots_ootb/effandfake1.pdf

**Comment by VinInn** [comment]
No is not this. Is the bug fix between pre2 and pre3
It is detailed in Kevin and Matti reports

**Comment by davidlange6** [comment]
maybe its worth correcting the PPD slides then..

**Pull request merged**